### PR TITLE
Switch atomic.Value to atomic.Pointer for spanProcessorStates

### PR DESCRIPTION
**Pull request by dmathieu**
This is a first look at switching `atomic.Value` to `atomic.Pointer`, so we have better type safety, and supposedly better readability.
I'm saying "supposedly" for better readability because for this specific case, it requires doing a lot of going from pointer to value, as a slice type can't be used from its pointer.

See #3918.

### CODE REVIEW

File: sdk/trace/provider.go [hunk 1]
```diff
@@ -180,7 +180,7 @@ func (p *TracerProvider) UnregisterSpanProcessor(sp SpanProcessor) {
 	if p.isShutdown {
 		return
 	}
-	old := p.spanProcessors.Load().(spanProcessorStates)
+	old := *(p.spanProcessors.Load())
```
Comment (pellared) [inline sdk/trace/provider.go:183]: how about extracting `*(p.spanProcessors.Load())` to a method e.g. named `getSpanProcessors` 😅

**Comment by pellared** [comment]
Is there any performance difference?

**Comment by dmathieu** [comment]
There is zero performance difference based on the existing benchmark tests.

```
pkg: go.opentelemetry.io/otel/sdk/trace
                                              │ bench-main  │            bench-branch            │
                                              │   sec/op    │   sec/op     vs base               │
SpanProcessor-10                                6.665µ ± 3%   6.501µ ± 2%  -2.47% (p=0.001 n=10)
SpanProcessorVerboseLogging-10                  6.543µ ± 2%   6.518µ ± 5%       ~ (p=0.684 n=10)
SpanLimits/AttributeValueLengthLimit-10         4.657µ ± 2%   4.621µ ± 2%       ~ (p=0.224 n=10)
SpanLimits/AttributeCountLimit-10               3.916µ ± 1%   3.873µ ± 1%  -1.10% (p=0.014 n=10)
SpanLimits/EventCountLimit-10                   4.228µ ± 1%   4.193µ ± 1%       ~ (p=0.143 n=10)
SpanLimits/LinkCountLimit-10                    4.218µ ± 2%   4.206µ ± 1%       ~ (p=0.897 n=10)
SpanLimits/AttributePerEventCountLimit-10       4.323µ ± 1%   4.326µ ± 1%       ~ (p=0.956 n=10)
SpanLimits/AttributePerLinkCountLimit-10        4.312µ ± 1%   4.303µ ± 1%       ~ (p=0.306 n=10)
SpanSetAttributesOverCapacity-10                1.430µ ± 1%   1.421µ ± 1%  -0.66% (p=0.000 n=10)
StartEndSpan/AlwaysSample-10                    389.7n ± 1%   384.8n ± 1%  -1.27% (p=0.000 n=10)
StartEndSpan/NeverSample-10                     175.2n ± 1%   172.4n ± 1%  -1.57% (p=0.000 n=10)
SpanWithAttributes_4/AlwaysSample-10            678.0n ± 1%   676.7n ± 1%       ~ (p=0.448 n=10)
SpanWithAttributes_4/NeverSample-10             257.2n ± 2%   255.9n ± 1%       ~ (p=0.160 n=10)
SpanWithAttributes_8/AlwaysSample-10            891.5n ± 2%   888.2n ± 1%       ~ (p=0.579 n=10)
SpanWithAttributes_8/NeverSample-10             318.9n ± 1%   320.3n ± 1%       ~ (p=0.403 n=10)
SpanWithAttributes_all/AlwaysSample-10          817.5n ± 1%   809.5n ± 3%       ~ (p=0.210 n=10)
SpanWithAttributes_all/NeverSample-10           275.6n ± 1%   275.8n ± 2%       ~ (p=0.956 n=10)
SpanWithAttributes_all_2x/AlwaysSample-10       1.135µ ± 1%   1.127µ ± 1%       ~ (p=0.270 n=10)
SpanWithAttributes_all_2x/NeverSample-10        359.5n ± 1%   359.8n ± 1%       ~ (p=0.699 n=10)
SpanWithEvents_4/AlwaysSample-10                876.8n ± 1%   857.5n ± 1%  -2.20% (p=0.001 n=10)
SpanWithEvents_4/NeverSample-10                 181.7n ± 1%   181.0n ± 1%       ~ (p=0.171 n=10)
SpanWithEvents_8/AlwaysSample-10                1.315µ ± 0%   1.311µ ± 1%       ~ (p=0.091 n=10)
SpanWithEvents_8/NeverSample-10                 188.4n ± 1%   187.2n ± 1%  -0.61% (p=0.005 n=10)
SpanWithEvents_WithStackTrace/AlwaysSample-10   557.1n ± 1%   546.9n ± 1%  -1.85% (p=0.000 n=10)
SpanWithEvents_WithStackTrace/NeverSample-10    198.5n ± 1%   197.4n ± 1%  -0.53% (p=0.024 n=10)
SpanWithEvents_WithTimestamp/AlwaysSample-10    540.5n ± 1%   532.2n ± 1%  -1.53% (p=0.000 n=10)
SpanWithEvents_WithTimestamp/NeverSample-10     230.8n ± 1%   228.9n ± 1%  -0.82% (p=0.001 n=10)
TraceID_DotString-10                            44.20n ± 0%   44.15n ± 0%       ~ (p=0.470 n=10)
SpanID_DotString-10                             34.76n ± 1%   34.61n ± 1%       ~ (p=0.055 n=10)
geomean                                         727.5n        722.4n       -0.71%

                                              │  bench-main  │             bench-branch              │
                                              │     B/op     │     B/op      vs base                 │
SpanProcessor-10                                8.266Ki ± 0%   8.266Ki ± 0%       ~ (p=1.000 n=10) ¹
SpanProcessorVerboseLogging-10                  8.266Ki ± 0%   8.266Ki ± 0%       ~ (p=1.000 n=10) ¹
SpanLimits/AttributeValueLengthLimit-10         10.39Ki ± 0%   10.39Ki ± 0%       ~ (p=1.000 n=10) ¹
SpanLimits/AttributeCountLimit-10               8.375Ki ± 0%   8.375Ki ± 0%       ~ (p=1.000 n=10) ¹
SpanLimits/EventCountLimit-10                   10.03Ki ± 0%   10.03Ki ± 0%       ~ (p=1.000 n=10) ¹
SpanLimits/LinkCountLimit-10                    10.03Ki ± 0%   10.03Ki ± 0%       ~ (p=1.000 n=10) ¹
SpanLimits/AttributePerEventCountLimit-10       10.25Ki ± 0%   10.25Ki ± 0%       ~ (p=1.000 n=10) ¹
SpanLimits/AttributePerLinkCountLimit-10        10.25Ki ± 0%   10.25Ki ± 0%       ~ (p=1.000 n=10) ¹
SpanSetAttributesOverCapacity-10                  496.0 ± 0%     496.0 ± 0%       ~ (p=1.000 n=10) ¹
StartEndSpan/AlwaysSample-10                      432.0 ± 0%     432.0 ± 0%       ~ (p=1.000 n=10) ¹
StartEndSpan/NeverSample-10                       128.0 ± 0%     128.0 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_4/AlwaysSample-10            1.109Ki ± 0%   1.109Ki ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_4/NeverSample-10               384.0 ± 0%     384.0 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_8/AlwaysSample-10            1.859Ki ± 0%   1.859Ki ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_8/NeverSample-10               640.0 ± 0%     640.0 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_all/AlwaysSample-10          1.672Ki ± 0%   1.672Ki ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_all/NeverSample-10             448.0 ± 0%     448.0 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_all_2x/AlwaysSample-10       2.984Ki ± 0%   2.984Ki ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_all_2x/NeverSample-10          768.0 ± 0%     768.0 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_4/AlwaysSample-10                  864.0 ± 0%     864.0 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_4/NeverSample-10                   128.0 ± 0%     128.0 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_8/AlwaysSample-10                1.281Ki ± 0%   1.281Ki ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_8/NeverSample-10                   128.0 ± 0%     128.0 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_WithStackTrace/AlwaysSample-10     544.0 ± 0%     544.0 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_WithStackTrace/NeverSample-10      144.0 ± 0%     144.0 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_WithTimestamp/AlwaysSample-10      568.0 ± 0%     568.0 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_WithTimestamp/NeverSample-10       168.0 ± 0%     168.0 ± 0%       ~ (p=1.000 n=10) ¹
geomean                                         1.203Ki        1.203Ki       +0.00%
¹ all samples are equal

                                              │ bench-main │            bench-branch             │
                                              │ allocs/op  │ allocs/op   vs base                 │
SpanProcessor-10                                36.00 ± 0%   36.00 ± 0%       ~ (p=1.000 n=10) ¹
SpanProcessorVerboseLogging-10                  36.00 ± 0%   36.00 ± 0%       ~ (p=1.000 n=10) ¹
SpanLimits/AttributeValueLengthLimit-10         63.00 ± 0%   63.00 ± 0%       ~ (p=1.000 n=10) ¹
SpanLimits/AttributeCountLimit-10               54.00 ± 0%   54.00 ± 0%       ~ (p=1.000 n=10) ¹
SpanLimits/EventCountLimit-10                   55.00 ± 0%   55.00 ± 0%       ~ (p=1.000 n=10) ¹
SpanLimits/LinkCountLimit-10                    55.00 ± 0%   55.00 ± 0%       ~ (p=1.000 n=10) ¹
SpanLimits/AttributePerEventCountLimit-10       58.00 ± 0%   58.00 ± 0%       ~ (p=1.000 n=10) ¹
SpanLimits/AttributePerLinkCountLimit-10        58.00 ± 0%   58.00 ± 0%       ~ (p=1.000 n=10) ¹
SpanSetAttributesOverCapacity-10                3.000 ± 0%   3.000 ± 0%       ~ (p=1.000 n=10) ¹
StartEndSpan/AlwaysSample-10                    2.000 ± 0%   2.000 ± 0%       ~ (p=1.000 n=10) ¹
StartEndSpan/NeverSample-10                     2.000 ± 0%   2.000 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_4/AlwaysSample-10            6.000 ± 0%   6.000 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_4/NeverSample-10             3.000 ± 0%   3.000 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_8/AlwaysSample-10            7.000 ± 0%   7.000 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_8/NeverSample-10             3.000 ± 0%   3.000 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_all/AlwaysSample-10          7.000 ± 0%   7.000 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_all/NeverSample-10           3.000 ± 0%   3.000 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_all_2x/AlwaysSample-10       8.000 ± 0%   8.000 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithAttributes_all_2x/NeverSample-10        3.000 ± 0%   3.000 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_4/AlwaysSample-10                9.000 ± 0%   9.000 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_4/NeverSample-10                 2.000 ± 0%   2.000 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_8/AlwaysSample-10                14.00 ± 0%   14.00 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_8/NeverSample-10                 2.000 ± 0%   2.000 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_WithStackTrace/AlwaysSample-10   5.000 ± 0%   5.000 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_WithStackTrace/NeverSample-10    3.000 ± 0%   3.000 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_WithTimestamp/AlwaysSample-10    6.000 ± 0%   6.000 ± 0%       ~ (p=1.000 n=10) ¹
SpanWithEvents_WithTimestamp/NeverSample-10     4.000 ± 0%   4.000 ± 0%       ~ (p=1.000 n=10) ¹
geomean                                         8.628        8.628       +0.00%
¹ all samples are equal
```

**Review by pellared — APPROVED** [review]
I am OK with the PR, but TBH I am not sure if it is worth addressing globally.

Personally, I see almost no value. Someone needs to write the code, then others need to review and a maintainer needs to merge 😉

However, for sure it would be OK to use `atomic.Pointer` in "new codebase" or during refactoring. It is good to know that there are no performance differences.

**Comment by dmathieu** [comment]
I agree, there is null value in doing this everywhere. I did it in the tracer as a quick one to start looking at the related issue. But I don't think we should apply the principle in other places.

**Review by hanyuancheung — APPROVED** [review]

**Pull request merged**